datasource/ethereum/src/block_stream.rs

```diff
@@ -1096,5 +1225,188 @@ fn create_log_filter_from_subgraph(manifest: &SubgraphManifest) -> EthereumLogFi
                     (contract_addr, event_sig)
                 })
         })
-        .collect::<EthereumLogFilter>()
+        .collect::<EthereumLogFilter>();
+    match log_filter.contract_address_and_event_sig_pairs.len() {
+        0 => return None,
+        _ => return Some(log_filter),
+    }
+}
+
+fn create_call_filter_from_subgraph(manifest: &SubgraphManifest) -> Option<EthereumCallFilter> {
+    let call_filter = manifest
+        .data_sources
+        .iter()
+        .filter(|data_source| {
+            data_source.source.address.is_some()
+        })
+        .flat_map(|data_source| {
+            let contract_addr = data_source.source.address;
+            data_source
+                .mapping
+                .call_handlers
+                .iter()
+                .map(move |call_handler| {
+                    let sig = keccak256(call_handler.function.as_bytes());
+                    (contract_addr.unwrap(), [sig[0], sig[1], sig[2], sig[3]])
+                })
+        })
+        .collect::<EthereumCallFilter>();
+    match call_filter.contract_addresses_function_signatures.len() {
+        0 => return None,
+        _ => return Some(call_filter),
+    }
+}
+
+fn create_block_filter_from_subgraph(manifest: &SubgraphManifest) -> Option<EthereumBlockFilter> {
+    let block_filter = manifest
+        .data_sources
+        .iter()
+        .filter(|data_source| {
+            data_source
+                .mapping
+                .block_handler
+                .is_some() && data_source.
+                source.
+                address.
+                is_some()
+        })
+        .map(|data_source| {
+            data_source.source.address.unwrap()
+        })
+        .collect::<EthereumBlockFilter>();
+    match block_filter.contract_addresses.len() {
+        0 => return None,
+        _ => return Some(block_filter),
+    }
+}
+
+fn parse_triggers(
+    log_filter_opt: Option<EthereumLogFilter>,
+    call_filter_opt: Option<EthereumCallFilter>,
+    block_filter_opt: Option<EthereumBlockFilter>,
+    descendant_block: &EthereumBlockWithCalls,
+) -> Result<Vec<EthereumTrigger>, Error> {
+    let mut triggers = Vec::new();
+    triggers.append(&mut parse_log_triggers(
```

Comment:
Isn't this basically
```rust
triggers.extend(parse_log_triggers(...));
```
? I know there's a difference but I'd think `extend` would be slightly more efficient and... prettier.

Reply:
My thought was that`append` avoids cloning the elements in the vector argument, will keep as is unless you object